### PR TITLE
More Code Examples from the Programming Phoenix book

### PR DESCRIPTION
**Pull request by nelsonic**
@iteles adding code examples for the Programming Phoenix book 
(_I've personally written and tested the examples and been opening issues along the way e.g #38 so other people can learn **faster**..._) 👍 

Fixes https://github.com/dwyl/learn-phoenix-framework/issues/35, https://github.com/dwyl/learn-phoenix-framework/issues/49 & https://github.com/dwyl/learn-phoenix-framework/issues/50

Please review/merge when you can thanks.

### CODE REVIEW

File: rumbl/priv/repo/seeds.exs [hunk 1]
```diff
@@ -9,3 +9,11 @@
 #
 # We recommend using the bang functions (`insert!`, `update!`
 # and so on) as they will fail if something goes wrong.
+
+alias Rumbl.Repo
+alias Rumbl.Category
+
+for category <- ~w(Action Drama Romance Comedy Sci-fi) do
+  Repo.insert!(%Category{name: category}) ||
+    Repo.insert!(%Category{name: category})
```
Comment (shouston3) [inline rumbl/priv/repo/seeds.exs]: I think these should be different, I have:

```
  Repo.get_by(Category, name: category) ||
    Repo.insert!(%Category{name: category})
```

instead

File: rumbl/priv/repo/migrations/20170218093549_create_category.exs [hunk 2]
```diff
@@ -0,0 +1,13 @@
+defmodule Rumbl.Repo.Migrations.CreateCategory do
+  use Ecto.Migration
+
+  def change do
+    create table(:categories) do
+      add :name, :string, null: false
+
+      timestamps() # function call parenthesis not in book!
```
Comment (shouston3) [inline rumbl/priv/repo/migrations/20170218093549_create_category.exs]: Nice 👍

File: rumbl/test/controllers/auth_test.exs [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+    assert conn.assigns.current_user.id == user.id
+  end
+
+  test "login with a not found user", %{conn: conn} do
+    assert {:error, :not_found, _conn} =
+      Auth.login_by_username_and_pass(conn, "me", "secret", repo: Repo)
+  end
+
+  test "login with password missmatch", %{conn: conn} do
+    _ = insert_user(username: "me", password: "secrete")
```
Comment (shouston3) [inline rumbl/test/controllers/auth_test.exs]: "secrete" typo

File: rumbl/test/models/user_test.exs [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+  test "changeset with valid attributes" do
+    changeset = User.changeset(%User{}, @valid_attrs)
+    assert changeset.valid?
+  end
+
+  # see: https://github.com/dwyl/learn-phoenix-framework/issues/41 #HelpWanted
+  # test "changeset with invalid attributes" do
+  #   changeset = User.changeset(%User{}, @invalid_attrs)
+  #   refute changeset.valid?
+  # end
```
Comment (shouston3) [inline rumbl/test/models/user_test.exs]: I also have these comments in mine

File: rumbl/web/controllers/video_controller.ex [hunk 5]
```diff
@@ -3,19 +3,28 @@ defmodule Rumbl.VideoController do
...
+
     render(conn, "new.html", changeset: changeset)
   end
 
-  def create(conn, %{"video" => video_params}) do
-    changeset = Video.changeset(%Video{}, video_params)
-
+  def create(conn, %{"video" => video_params}, user) do
+    IO.inspect video_params
+    IO.inspect user
```
Comment (shouston3) [inline rumbl/web/controllers/video_controller.ex]: Perhaps remove these `IO.inspects`

File: rumbl/web/controllers/video_controller.ex [hunk 6]
```diff
@@ -3,19 +3,28 @@ defmodule Rumbl.VideoController do
...
+    changeset =
+    user
+      |> build_assoc(:videos)
+      |> Video.changeset(video_params)
+    IO.inspect changeset
```
Comment (shouston3) [inline rumbl/web/controllers/video_controller.ex]: `IO.inspect`

File: rumbl/web/models/video.ex [hunk 7]
```diff
@@ -6,16 +6,23 @@ defmodule Rumbl.Video do
     field :title, :string
     field :description, :string
     belongs_to :user, Rumbl.User
+    belongs_to :category, Rumbl.Category
 
-    timestamps()
+    timestamps() # invocation parenthesis not in book
```
Comment (shouston3) [inline rumbl/web/models/video.ex]: 👍

File: rumbl/priv/repo/seeds.exs (hunk 1 above)
Comment (nelsonic) [inline rumbl/priv/repo/seeds.exs]: @shouston3 good shout. please feel free to edit the file in PR. 👍 
`else` I will fix tomorrow morning. 😉

File: rumbl/test/controllers/auth_test.exs (hunk 3 above)
Comment (nelsonic) [inline rumbl/test/controllers/auth_test.exs]: yeah, worth fixing. thanks. 👍

File: rumbl/web/controllers/video_controller.ex (hunk 6 above)
Comment (nelsonic) [inline rumbl/web/controllers/video_controller.ex]: Will remove. in next commit. 👍

File: rumbl/web/models/video.ex (hunk 7 above)
Comment (des-des) [inline rumbl/web/models/video.ex]: Ugh this has been confusing us with warnings

Comment (nelsonic) [inline rumbl/web/models/video.ex]: @des-des are you running the latest version of Phoenix?

Comment (des-des) [inline rumbl/web/models/video.ex]: @nelsonic yeah 1.2.1

File: rumbl/web/static/js/player.js [hunk 8]
```diff
@@ -0,0 +1,29 @@
+let Player = {
```
Comment (des-des) [inline rumbl/web/static/js/player.js]: JS -> Elm?!

Comment (nelsonic) [inline rumbl/web/static/js/player.js]: @des-des this example is from "The Book" ...
I don't _intend_ to keep the ES6 after I've finished writing out the code. 😉

Comment (des-des) [inline rumbl/web/static/js/player.js]: hehe

**Comment by nelsonic** [comment]
Totes works!
image

**Comment by nelsonic** [comment]
@iteles I've addressed the review comments from @shouston3  & @des-des 
and added Travis-CI to confirm that the demo app is working.
Please take a look at the code when you have time and merge if ok. 
thanks! 👍

File: rumbl/web/static/js/player.js [hunk 9]
```diff
@@ -1,6 +1,7 @@
-let Player = {
-  player: null,
-
+let Player = {  // JS works for small scripts like this
```
Comment (des-des) [inline rumbl/web/static/js/player.js]: :heart:

Comment (iteles) [inline rumbl/web/static/js/player.js]: Also loving this pragmatism 👍

**Pull request merged**